### DISCUSSION TITLE
Loading oqsprovider fails under Windows

### DISCUSSION THREAD

**Discussion by VeronikaNguyen** [openssl/openssl#21903] [Q&A] [answered]
I am trying to load the oqsprovider via code under Windows since loading the provider over the openssl config is not an option for me. 
The `oqsprovider.dll` file is located in `D:\oqsprovider`. When running the following commands
```  
OSSL_PROVIDER_set_default_search_path(NULL, "D:\\oqsprovider");
auto path = OSSL_PROVIDER_get0_default_search_path(NULL);
auto loaded = OSSL_PROVIDER_load(NULL, "oqsprovider");
```
`path` is set correctly, however, `loaded` is `NULL`, which indicates that the provider was not successfully loaded.

When printing out the error with 
```
  int err = ERR_get_error();
  char buf[256];
  ERR_error_string_n(err, buf, sizeof(buf));
```
it says `error:078C0105:common libcrypto routines::init fail`. 

Am I missing something here? I am grateful for every kind of input.

Note that I asked the same question here: https://github.com/open-quantum-safe/oqs-provider/issues/237

**Comment by mattcaswell** [comment]
> error:078C0105:common libcrypto routines::init fail

There are two possible locations in the code base where this error can be emitted. One occurs when something has called `OPENSSL_init_crypto()` after OpenSSL has already been de-initialised. That doesn't seem likely here given your description of the problem.

Most likely it is the other location which is here:

https://github.com/openssl/openssl/blob/94a48cc9e08820d0b96904ac8c646eefde61e181/crypto/provider_core.c#L922-L927

This is the code that calls the provider's `OSSL_provider_init` function. To get here we have successfully found and loaded the provider, but when we have tried to initialise it the provider has returned an error.

To confirm this you might want to call `ERR_get_error_all()` instead of `ERR_get_error()`:

https://www.openssl.org/docs/man3.1/man3/ERR_get_error_all.html

This is the same as `ERR_get_error()` but additionally gives you more information about the error such as the OpenSSL source file and line number that raised it as well as an additional "data" string - which in this case should tell you the name of the provider that failed to initialise. I'd suggest trying that function and printing the additional information to see if it gives us any further clues as to what is going wrong here.

**Reply by VeronikaNguyen** [reply]
Thank you very much for the helpful input here. Your assumptions are confirmed by `ERR_get_error_all()` which has the following output:

```
OQS PROV: Impossible error: NID unregistered for dilithium2.
file: crypto\provider_core.c
line: 937
func: provider_init
data: name=oqsprovider
flags: 3
```

**Reply by mattcaswell** [reply]
That looks like an OQS Provider problem.

@baentsch

**Reply by baentsch** [reply]
> That looks like an OQS Provider problem.

Hmm... in this case this it output as `OBJ_sn2nid` returned 0 _after successfully_ going through `c_obj_create` (`core_obj_create`) -- always with the same name ("dilithium2" in this case). Can you think of a reason how this could happen, @mattcaswell ? Why only on Windows (see https://github.com/open-quantum-safe/oqs-provider/issues/237#issuecomment-1699383466)? Why only for @VeronikaNguyen ? `oqsprovider` CI passes OK on Windows.... Do you see something suspicious in [here](https://github.com/open-quantum-safe/oqs-provider/blob/b6da1e97b7180a779981d1b0bfa0ab2cd723066d/oqsprov/oqsprov.c#L593-L619)?

The only statement in the documentation that makes me pause is 

> The core_obj_create() and core_obj_add_sigid() functions were not thread safe in OpenSSL 3.0.

Is this resolved for good also under Windows under all circumstances? @VeronikaNguyen: Are you using OpenSSL 3.0 or something more recent?

**Reply by VeronikaNguyen** [reply]
@baentsch I am using something more recent from the master branch of openssl, OpenSSL 3.2.0-dev (from May 2023).

**Reply by mattcaswell** [reply]
> in this case this it output as OBJ_sn2nid returned 0 after successfully going through c_obj_create (core_obj_create) -- always with the same name ("dilithium2" in this case). Can you think of a reason how this could happen, @mattcaswell?

Yes. This could happen in some circumstances, i.e. if the oqsprovider and the application are linked against *different* instances of OpenSSL. This is the reason why the `c_obj_create` callback exists in the first place. It is calling the *application's* version of `OBJ_create` (aka the "core") and not the oqsprovider's version. I think this is what @slontis is referring to in his [comment](https://github.com/openssl/openssl/discussions/21903#discussioncomment-6882083)

**Reply by baentsch** [reply]
BIG Aha! Thanks @mattcaswell, I wasn't aware of that : This almost certainly is the case here (AFAIK @VeronikaNguyen built a local binary of OpenSSL 3.2.0-dev and uses an `oqsprovider` binary built against "some" version of OpenSSL3.2.0-dev). 

So the question for me then is: How can this be avoided (short of removing the use of NIDs throughout the code base of OpenSSL -- or at least within providers?)? Make available an "object query" callback analogous to `c_obj_create`? Or would it be a worthwhile exercise to go through "oqsprovider" to (try to) eliminate all NID uses (if possible at all)?

**Reply by t8m** [reply]
The provider simply should not call OBJ_sn2nid().  They have to believe that the c_obj_create works fine.

**Reply by t8m** [reply]
> Or would it be a worthwhile exercise to go through "oqsprovider" to (try to) eliminate all NID uses (if possible at all)?

Yes, it cannot use NIDs. Or, actually it can, but then it has to create the objects also internally with OBJ_create (and ignore the failure if the object is duplicate because it was already created by core_obj_create). And of course it cannot depend on the NID as the oqsprovider sees it is the same as the NID that the application sees.

**Reply by baentsch** [reply]
Thanks for that additional guidance, @t8m . At "first glance" it seems `oqsprovider` needs NIDs solely during encoding/decoding for interaction with X509, e.g., calling `OBJ_nid2obj`, `PKCS8_pkey_set0`, `X509_PUBKEY_set0_param`. Do you see a way around this? Could `oqsprovider` make use of `OBJ_txt2obj` instead of `OBJ_nid2obj` (always using the algorithm names it registered with `c_obj_create` in the first place)? But how to handle PKCS8 and X509_PUBKEYs without NIDs?

OR would it indeed be permissible to use `OBJ_create` (instead of `core_obj_create`) and keep using the above X509 APIs with NIDs? Would that avoid the issue @VeronikaNguyen ran into?

**Reply by t8m** [reply]
I'll need to study how oqsprovider uses NIDs internally (i.e. not the application that is calling the oqsprovider but the provider itself inside its code).

**Reply by mattcaswell** [reply]
> would it indeed be permissible to use OBJ_create (instead of core_obj_create)

You could call both. The second one may fail if both core and oqsprovider are using the same OpenSSL instance (but you can just ignore the failure). core and oqsprovider may be using different NIDs for the same thing - but that shouldn't matter.

**Reply by t8m** [reply]
As first quick fix you can just call OBJ_create() in addition to the c_obj_create() and ignore an eventual failure. This will ensure that you will be still able to use nids internally. Hopefully the oqsprovider code does not depend on the nids being directly passed to the application - it should not as there are no provider APIs that would accept or return NIDs.

Eventually you should get rid of NIDs completely. For the application working, it is sufficient to call c_obj_create() and that will care of the algorithm having a NID on the application side, but on the provider side I do not see a reason why you would need to use them - with some code rewriting as I see there are quite a few places where NIDs are being used.

**Reply by baentsch** [reply]
> You could call both. The second one may fail if both core and oqsprovider are using the same OpenSSL instance (but you can just ignore the failure). core and oqsprovider may be using different NIDs for the same thing - but that shouldn't matter.

Calling one is absolutely fine for me -- I'd actually prefer `OBJ_create` but thought providers _have to_ call the `c_create` callback for some (bookkeeping?) reason.

**Reply by baentsch** [reply]
> as I see there are quite a few places where NIDs are being used.

Yup - those are the thing. I didn't find a way around them to make use of X509/PKCS8 APIs. If there'd be some documentation how to do that (without NIDs), I'd be eager to learn about that.

**Reply by mattcaswell** [reply]
> Calling one is absolutely fine for me -- I'd actually prefer OBJ_create but thought providers have to call the c_create callback for some (bookkeeping?) reason.

If you don't call `core_obj_create` then the NIDs may not be visible to core. If you don't call `OBJ_create` then the NIDs may not be visible to oqsprovider. If they need to be visible to both then you will have to call both of them. In cases where there is only one instance of OpenSSL (i.e. the usual case) then this isn't necessary. But in cases like this one it is needed.

**Reply by t8m** [reply]
> Yup - those are the thing. I didn't find a way around them to make use of X509/PKCS8 APIs. If there'd be some documentation how to do that (without NIDs), I'd be eager to learn about that.

It might be that for now you'll have to use NIDs. Eventually the OpenSSL APIs should get rid of NIDs completely and the functions using NIDs will be deprecated. But yeah for now just call both create calls and ignore eventual failure of the OBJ_create() (that should be the second call).

**Reply by baentsch** [reply]
So we'll call both. No problem. If this solves such issues going forward, I'm all for it. @VeronikaNguyen stay tuned for an update in https://github.com/open-quantum-safe/oqs-provider/issues/237

**Reply by baentsch** [reply]
Update: This doesn't work. OpenSSL hangs when adding an additional `OBJ_create`. See [CI log](https://github.com/open-quantum-safe/oqs-provider/actions/runs/6056373314/job/16436709497?pr=243) for https://github.com/open-quantum-safe/oqs-provider/pull/243. Will report when finding the root cause for this (some lock somewhere?).

**Reply by baentsch** [reply]
Update 2: Locking error only in OpenSSL 3.1.0. Workaround for workaround in https://github.com/open-quantum-safe/oqs-provider/pull/243/commits/696d944d31c1b2ec58c18cf939472b0393a37f6c.

**Comment by slontis** [comment, marked as answer]
I assume there is a very good reason that  NID's still need to be used here?
This seems fishy to me.. The c_obj_create() is there because it is resolving a function that belongs to core, and it is passed to your provider so that it can use the call in the oqs provider. Calling OBJ_sn2nid() is clearing doing something different to this. Is this accessing core or does this symbol come from the oqs provider (potentially using its own copy).

**Comment by baentsch** [comment]
> I assume there is a very good reason that NID's still need to be used here? This seems fishy to me.. The c_obj_create() is there because it is resolving a function that belongs to core, and it is passed to your provider so that it can use the call in the oqs provider. Calling OBJ_sn2nid() is clearing doing something different to this. Is this accessing core or does this symbol come from the oqs provider (potentially using its own copy).

If NIDs are not registered (via `core_obj_create`), AFAIK the complete signature & X509 logic of OpenSSL fails. In this particular location calling (the core function) `OBJ_sn2nid` is just a sanity check that they got correctly registered. In other places of the provider the calls to that function are there to support encoding/decoding, for example. And everything works OK this way in UNIX and our Windows CI environment.

FWIW, this is going back to https://github.com/openssl/openssl/issues/10512#issuecomment-1138841073 . If something changed since then please let me know.

**Reply by baentsch** [reply]
FWIW, here's another take on NIDs: https://github.com/openssl/openssl/issues/20594 May I ask what the current plan is about (keeping/getting rid of) them?

**Reply by t8m** [reply]
This will happen at some point, but not for 3.2.
